### PR TITLE
exec sync support

### PR DESCRIPTION
**Pull request by rphillips**
- [X] ~Create a capnproto API to stream stdout/stderr from conmon-rs~
- [X] actually implement exec
- [x] race for some reason
- [x] migrate to tokio spawn
- [x] refactor the pid watcher

### CODE REVIEW

File: conmon-rs/common/proto/conmon.capnp [hunk 1]
```diff
@@ -19,4 +23,33 @@ interface Conmon {
...
+
+    enum ExecChunkType {
+        stdout @0;
+        stderr @1;
+    }
+
+    interface ExecCallback {
+        sendChunk @0 (type :ExecChunkType, chunk :Data);
+        done @1 ();
+    }
```
Comment (saschagrunert) [inline conmon-rs/common/proto/conmon.capnp:52]: I was thinking if it would make sense to merge `done` into `sendChunk` and pass the exit code as another parameter. But I guess having two methods is fine as well.

**Comment by haircommander** [comment]
while this is a good start, I think it may be better to focus on ExecSync requests--ones that wait for the entirety of the output from the exec request. this is currently what CRI-O uses conmon for, and it will be simpler than a stream

File: conmon-rs/server/src/rpc.rs [hunk 2]
```diff
@@ -69,6 +69,34 @@ impl conmon::Server for Server {
...
+            let child_reaper = Arc::clone(self.reaper());
+            let command = self.generate_exec_sync_args(&params).unwrap();
+            let (exit_status, stdout, stderr) =
+                pry_err!(child_reaper.exec_sync(self.config().runtime(), command));
+            let mut resp = results.get().init_response();
+            resp.set_exit_code(exit_status.code().unwrap());
+            resp.set_stdout(stdout.as_str());
+            resp.set_stderr(stderr.as_str());
+        } else {
+            // container does not exist
```
Comment (haircommander) [inline conmon-rs/server/src/rpc.rs:93]: in this situation we should error. I would reverse the blocks and return with an error if None here

File: conmon-rs/server/src/child_reaper.rs [hunk 3]
```diff
@@ -59,40 +60,77 @@ impl ChildReaper {
...
+        let delay = tokio::time::sleep(tokio::time::Duration::from_secs(timeout as u64));
+        tokio::pin!(delay);
+        tokio::select! {
+            _ = &mut delay => {
+                debug!("timeout");
+                return Err(format_err!("timeout error"))
+            },
+            status = child.wait_with_output() => {
+                 let output = status.expect("blah");
+                 let stdout = String::from_utf8(output.stdout)?;
```
Comment (cgwalters) [inline conmon-rs/server/src/child_reaper.rs:101]: I think we don't want to error out if the child output invalid UTF-8 somehow.  I'd use [from_utf8_lossy](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy) instead.

(And what's better, we avoid duplicating the buffer in the happy path)

However, since this function returns owned values, we probably want to change it to return the raw `Vec<u8>`, and do the utf8 bits in the callers.

File: conmon-rs/server/src/rpc.rs [hunk 4]
```diff
@@ -48,29 +49,80 @@ impl conmon::Server for Server {
...
+        let runtime = self.config.runtime().clone();
+        debug!(
+            "Got exec sync container request for id {} with timeout {} : {}",
+            id,
+            timeout,
+            command.join(" ")
+        );
+        let children = self.children.read().unwrap();
+        let child = children.get(id);
+        if child.is_none() {
```
Comment (cgwalters) [inline conmon-rs/server/src/rpc.rs:108]: There's a bit more verbose pattern for this
```
let child = if let Some(c) = child {
    c
} else {
    let mut resp = results.get().init_response();
    resp.set_exit_code(-1);
    return Promise::ok(());
};
```
And then below we can drop the `unwrap()`.

In the future https://rust-lang.github.io/rfcs/3137-let-else.html will help this.

File: conmon-rs/server/src/rpc.rs [hunk 5]
```diff
@@ -48,29 +49,80 @@ impl conmon::Server for Server {
...
+        debug!("found child with id {}", child.unwrap().id);
+        let child_reaper = Arc::clone(self.reaper());
+        Promise::from_future(async move {
+            let (exit_status, stdout, stderr) = child_reaper
+                .exec_sync(&runtime, command, timeout)
+                .await
+                .map_err(|e| Error::new(ErrorKind::Other, format!("Error {}", e)))?;
+            let mut resp = results.get().init_response();
+            resp.set_exit_code(exit_status.code().unwrap());
+            resp.set_stdout(stdout.as_str());
```
Comment (cgwalters) [inline conmon-rs/server/src/rpc.rs:122]: If `exec_sync` returned `Vec<u8>` this is where we could do `from_utf8_lossy()`.

File: conmon-rs/server/src/rpc.rs (hunk 4 above)
Comment (rphillips) [reply conmon-rs/server/src/rpc.rs]: thanks, incorporated this in!

File: conmon-rs/server/src/rpc.rs (hunk 5 above)
Comment (rphillips) [reply conmon-rs/server/src/rpc.rs]: thanks. fixed :)

File: conmon-rs/server/src/child_reaper.rs [hunk 6]
```diff
@@ -7,133 +7,88 @@ use log::{debug, error};
...
+        args: Vec<String>,
+        timeout: i32,
+    ) -> Result<Output> {
+        let child = tokio::process::Command::new(command)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .kill_on_drop(true)
+            .args(args)
+            .spawn()
+            .map_err(|e| format_err!("spawn child process: {}", e))?;
```
Comment (saschagrunert) [inline conmon-rs/server/src/child_reaper.rs:42]: Let's replace those with the `anyhow::Context`:
```suggestion
            .context("spawn child process")?;
```

File: conmon-rs/server/src/child_reaper.rs [hunk 7]
```diff
@@ -7,133 +7,88 @@ use log::{debug, error};
...
+
+        let delay = tokio::time::sleep(tokio::time::Duration::from_secs(timeout as u64));
+        tokio::pin!(delay);
+        tokio::select! {
+            _ = &mut delay => {
+                debug!("timeout");
+                return Err(format_err!("timeout error"))
```
Comment (saschagrunert) [inline conmon-rs/server/src/child_reaper.rs:49]:
```suggestion
                bail!("timeout error")
```

File: conmon-rs/server/src/child_reaper.rs [hunk 8]
```diff
@@ -7,133 +7,88 @@ use log::{debug, error};
...
+            },
+            status = child.wait_with_output() => {
+                 let output = status.expect("status expected");
+                 Ok(output)
```
Comment (saschagrunert) [inline conmon-rs/server/src/child_reaper.rs:53]:
```suggestion
                 Ok(status?)
```

File: conmon-rs/server/src/lib.rs [hunk 9]
```diff
@@ -236,4 +246,28 @@ impl Server {
...
+    fn generate_exec_sync_args(
+        &self,
+        params: &conmon::ExecSyncContainerParams,
+    ) -> Result<Vec<String>> {
+        let req = params.get()?.get_request()?;
+        let id = req.get_id()?.to_string();
+        let command = req.get_command()?;
+        let runtime_root = self.config().runtime_root();
+
+        let mut args: Vec<String> = vec![];
```
Comment (saschagrunert) [inline conmon-rs/server/src/lib.rs:260]: nit:
```suggestion
        let mut args = vec![];
```

File: conmon-rs/server/src/lib.rs [hunk 10]
```diff
@@ -236,4 +246,28 @@ impl Server {
...
+        if let Some(rr) = runtime_root {
+            args.push(format!("--root={}", rr.display()));
+        }
+        args.push("exec".to_string());
+        args.push(id);
+        for value in command.iter() {
+            args.push(value.unwrap().to_string());
```
Comment (saschagrunert) [inline conmon-rs/server/src/lib.rs:267]:
```suggestion
            args.push(value?.to_string());
```

File: conmon-rs/server/src/lib.rs (hunk 9 above)
Comment (rphillips) [reply conmon-rs/server/src/lib.rs]: fixed!

File: conmon-rs/server/src/child_reaper.rs (hunk 8 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs]: fixed!

File: conmon-rs/server/src/child_reaper.rs (hunk 7 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs]: fixed!

File: conmon-rs/server/src/child_reaper.rs (hunk 6 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs:74]: fixed!

**Comment by rphillips** [comment]
Need to figure out the stdout/stderr streams... and timeouts.

Transitioned to using "-d" for detach on the exec.

File: conmon-rs/server/src/child_reaper.rs [hunk 11]
```diff
@@ -1,169 +1,148 @@
...
-                }
-            }
-        });
-        Ok(())
+impl ChildReaper {
+    pub fn get(&self, id: String) -> Result<ReapableChild> {
+        let locked_grandchildren = Arc::clone(&self.grandchildren);
+        let lock = locked_grandchildren.lock().unwrap();
+        let r = lock.get(&id).context("")?.clone();
+        drop(lock);
```
Comment (haircommander) [inline conmon-rs/server/src/child_reaper.rs:24]: do we need to manually call drop here? I would expect it to go out of scope after the Ok

**Comment by haircommander** [comment]
could you separate the commits to be smaller pieces of work?

File: conmon-rs/server/src/child_reaper.rs (hunk 11 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs:24]: yes, the drop needs to be called to allow the borrow checker to succeed.

**Comment by rphillips** [comment]
This is ready for review.

There are a couple pending TODOs. The console library needs a refactor to propagate up the stdout/stderr of the grandchild. The secondary TODO is to return the grandchild's exit code.

I think we should merge this and then tackle the TODOs in parallel.

File: conmon-rs/common/proto/conmon.capnp [hunk 12]
```diff
@@ -23,4 +27,20 @@ interface Conmon {
     }
 
     createContainer @1 (request: CreateContainerRequest) -> (response: CreateContainerResponse);
-}
+
```
Comment (haircommander) [inline conmon-rs/common/proto/conmon.capnp:30]: i think these changes should go in the second commit

File: conmon-rs/server/src/lib.rs [hunk 13]
```diff
@@ -236,4 +235,36 @@ impl Server {
...
+        }
+        args.push(format!("--pid-file={}", pidfile.display()));
+        args.push(id);
+        for value in command.iter() {
+            args.push(value?.to_string());
+        }
+
+        debug!("Exec args {:?}", args.join(" "));
+        Ok(args)
+    }
```
Comment (haircommander) [inline conmon-rs/server/src/lib.rs:269]: as well as these (and any specifically exec related pieces)

File: conmon-rs/server/src/child_reaper.rs [hunk 14]
```diff
@@ -1,169 +1,148 @@
...
+                    if err != nix::errno::Errno::ECHILD {
+                        error!("caught error in reading for sigchld {}", err);
+                    }
+                }
+            };
+            if tx.send(()).is_err() {
+                error!("the receiver dropped the watch")
+            }
+        });
+        rx
```
Comment (haircommander) [inline conmon-rs/server/src/child_reaper.rs:141]: whereas the refactoring of child reaping and pid watching should be in a separate commit describing the reasoning

File: conmon-rs/server/src/rpc.rs [hunk 15]
```diff
@@ -48,29 +48,82 @@ impl conmon::Server for Server {
...
+            // TODO FIXME why convert?
+            results
+                .get()
+                .init_response()
+                .set_container_pid(grandchild_pid as u32);
+            Ok(())
+        })
+    }
+
+    fn exec_sync_container(
```
Comment (haircommander) [inline conmon-rs/server/src/rpc.rs:76]: same here--second commit adding exec please

File: pkg/client/client.go [hunk 16]
```diff
@@ -166,6 +166,68 @@ func (c *ConmonClient) CreateContainer(ctx context.Context, cfg *CreateContainer
 	return response.ContainerPid(), nil
 }
 
+type ExecContainerResult struct {
```
Comment (haircommander) [inline pkg/client/client.go:169]: client update could even go in another separate commit--but that's up to you. it could also go with the exec commit if you'd like

**Review by haircommander — CHANGES_REQUESTED** [review]
:tada: great work! I have described what I'm looking for in commit structure a bit more thoroughly. My first glance LGTM, once this is reorganized i'll do another more in-depth review

File: conmon-rs/common/proto/conmon.capnp (hunk 12 above)
Comment (rphillips) [reply conmon-rs/common/proto/conmon.capnp:30]: hmm. the second commit is just the regeneration of the files.

File: conmon-rs/server/src/child_reaper.rs [hunk 17]
```diff
@@ -1,169 +1,148 @@
...
+                Ok(status) => {
+                    if let WaitStatus::Exited(_, exit_status) = status {
+                        let _ = write_to_exit_paths(exit_status, &exit_paths);
+                    }
+                }
+                Err(err) => {
+                    if err != nix::errno::Errno::ECHILD {
+                        error!("caught error in reading for sigchld {}", err);
+                    }
+                }
```
Comment (haircommander) [inline conmon-rs/server/src/child_reaper.rs:135]: nit: we may want a TODO here. Part of me thinks we may want to write to the exit file anyway so the caller knows the container exited.

nit2: it's possible we want to retry in this case. we'll have to stress conmon more to figure it out

File: conmon-rs/server/src/child_reaper.rs [hunk 18]
```diff
@@ -1,169 +1,148 @@
...
-        if let Some(old) = map.insert(child.pid, reapable_grandchild) {
-            bail!("Repeat PID for container {} found", old.id);
-        }
+        let reapable_grandchild = ReapableChild::from_child(&child);
+        let killed_channel = reapable_grandchild.watch();
+        map.insert(child.id, reapable_grandchild);
+        let cleanup_grandchildren = locked_grandchildren.clone();
+        let pid = child.pid;
+        tokio::task::spawn(async move {
+            killed_channel.await.expect("no error on channel");
```
Comment (haircommander) [inline conmon-rs/server/src/child_reaper.rs:72]: wouldn't calling expect("no error on channel") mean we print an error saying "no error on channel" in cases where there was an error on the channel?

File: conmon-rs/server/src/child_reaper.rs [hunk 19]
```diff
@@ -1,169 +1,148 @@
...
+            if let Err(e) = Self::forget_grandchild(&cleanup_grandchildren, pid) {
```
Comment (haircommander) [inline conmon-rs/server/src/child_reaper.rs:73]: also, if we're waiting on the channel here to forget the granchild, why don't we just call waitpid within this task, rather than needing to sync between two tasks?

**Review by haircommander — APPROVED** [review]
A couple of more questions. I don't think any are blocking, but I wanted to get them out there. Being added as TODOs would be fine with me

Otherwise, LGTM

File: conmon-rs/server/src/child_reaper.rs (hunk 18 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs:72]: hah. yeah... I'll fix that :)

File: conmon-rs/server/src/child_reaper.rs (hunk 19 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs:73]: since this is the 'sync' function I am thinking we want to wait on the async task to complete. Ideally, we want less code and usually the 'sync' version is just a block on the async version.

File: conmon-rs/server/src/child_reaper.rs (hunk 17 above)
Comment (rphillips) [reply conmon-rs/server/src/child_reaper.rs]: fixed.

**Review by saschagrunert — APPROVED** [review]
Great work, let's follow-up on the open discussion items and iterate.
